### PR TITLE
Refactor some constraint into GenericConstraint

### PR DESCRIPTION
**Pull request by blegat**
This is a first step towards allowing to create constraints with custom MOI vector sets.

### CODE REVIEW

**Comment by odow** [comment]
Let me know when you'd like me to take a look

**Comment by odow** [comment]
Why would this be breaking for #589?

**Comment by blegat** [comment]
`PositiveSemidefiniteConeConstraint` etc... disappear and the sign of the dual of `<=` constraint is swapped

**Comment by odow** [comment]
> PositiveSemidefiniteConeConstraint etc... disappear 

We could keep the constant as an alias.

> and the sign of the dual of <= constraint is swapped

This needs a much bigger discussion. We could change it without this PR, but it's horribly breaking because it silently swaps the sign.

I don't know how many people rely on this though.

**Comment by odow** [comment]
> We can also avoid breaking this by using GenericConstraint{MOI.Nonnegatives} for both cases

:+1: at least for this PR.

> you seem to agree in view of the comment you wrote in the code

I agree that we should fix it. But I don't know how to do it in a way that doesn't trip people up. At the very least, it needs to be a single PR just for that change.

**Comment by ericphanson** [comment]
I think it would be good to give the expected sign, and I'm not sure so many folks are using the existing functionality; there's been no issues about the sign at least. Maybe we could tag a v0.15.5 with a warning in `getproperty` for the field `dual` that the sign is changing, so anyone accessing it will be warned? Or maybe the release notes are enough.

**Comment by blegat** [comment]
We are already changing the sign for the PSD cone:
https://github.com/jump-dev/Convex.jl/blob/7eb1726f4b77a9c598f783a80327c4d5fc46dd08/src/constraints/PositiveSemidefiniteConeConstraint.jl#L82-L84
so it's maybe consistent to consider that `a <= b` is `b - a in Nonnegatives()` as well. In JuMP, we avoid it by requiring the cone to be given explicitly for conic inequalities so rewriting  `@constraint(model, a <= b, PSDCone())` into `b - a in PSDCone()` shouldn't be confusing or surprising.

If we want to be on the safe side we could also simply throw a warning in v0.16 when someone gets a dual with a `<=` inequality and then in `v0.17` we can add `GenericConstraint{MOI.Nonpositives}` constraints.

**Comment by odow** [comment]
I've changed the sign in #593

**Comment by odow** [comment]
I think we should consider releasing v0.16 without this PR. 

There are a lot of good changes sitting on `master`, and this is a nice-to-have, but not a necessity. It's also not obvious how you could write things in a generic way that works for arbitrary sets. You'll probably need specialized code for each set anyway. So we could always add this as a feature addition for other sets in MOI, and only change things over in the 1.0 release.

**Comment by blegat** [comment]
>  So we could always add this as a feature addition for other sets in MOI, and only change things over in the 1.0 release.

I don't know how far we are from v1.0, I think this will be a v0.17 probably. Since v0.16 is going to be breaking, and v0.17 as well, it just think it might be best to break once to avoid people having to change back and forth.

> It's also not obvious how you could write things in a generic way that works for arbitrary sets.

Not much need to be done it seems. I'm planning to add default fallbacks in a follow-up PR. `head` could just print the set type, `is_feasible` could use MathOptSetDistances, `set_with_size` could use `MOI.Utilities.set_with_dimension` if the size is `(n, 1)` and error if there is more than 1 column. So the only custom method that would be needed for every set is `vexity`.

**Comment by odow** [comment]
> Not much need to be done it seems. 

There are still lots of test errors?

File: src/constraints/GreaterThanConstraint.jl [hunk 1]
```diff
@@ -1,69 +1,49 @@
...
-    end
+function set_with_size(::Type{MOI.Nonnegatives}, sz::Tuple{Int,Int})
+    return MOI.Nonnegatives(prod(sz))
 end
 
-head(io::IO, ::GreaterThanConstraint) = print(io, "≥")
+head(io::IO, ::MOI.Nonnegatives) = print(io, "≥")
+
+function is_feasible(f, ::MOI.Nonnegatives, tol)
+    return all(f .>= tol)
```
Comment (odow) [inline src/constraints/GreaterThanConstraint.jl:8]:
```suggestion
    return all(f .>= -tol)
```

File: src/constraints/LessThanConstraint.jl [hunk 2]
```diff
@@ -1,67 +1,25 @@
...
-    end
+function set_with_size(::Type{MOI.Nonpositives}, sz::Tuple{Int,Int})
+    return MOI.Nonpositives(prod(sz))
 end
 
-head(io::IO, ::LessThanConstraint) = print(io, "≤")
+head(io::IO, ::MOI.Nonpositives) = print(io, "≤")
+
+function is_feasible(f, ::MOI.Nonpositives, tol)
+    return all(f .<= -tol)
```
Comment (odow) [inline src/constraints/LessThanConstraint.jl:8]:
```suggestion
    return all(f .<= tol)
```

**Comment by odow** [comment]
Okay, I've added `Zeros` and `SecondOrderCone` support.

I've left the existing `ExponentialConeConstraint`, `GeomMeanEpiCone`, `GeomMeanHypoCone`, and `RelativeEntropyEpiCone` connstraints alone because they do something other than a trivial MOI constraint. (`ExponentialConeConstraint` in particular is a broadcast over arrays, rather than a single `MOI.ExponentialCone` constraint.)

**Review by odow — APPROVED** [review]

**Comment by blegat** [comment]
> There are still lots of test errors?

I meant that not much need to be done "per set" once we have `GenericConstraint` working

**Pull request merged**